lib/list.gi

```diff
@@ -3674,17 +3674,17 @@ end );
 ##
 InstallGlobalFunction( ListWithIdenticalEntries, function( n, obj )
     local list, i;
-    if IsChar(obj) then
-      list := "";
-      for i in [ 1 .. n ] do
-        list[i]:= obj;
-      od;
+
+    if not IsInt(n) or n < 0 then
+      ErrorNoReturn("the first argument <n> must be a non-negative integer");
+    elif IsChar(obj) then
+      list := EmptyString(n);
     else
-      list:= [];
-      for i in [ n, n-1 .. 1 ] do
-        list[i]:= obj;
-      od;
```

Comment:
Note that assigning the elements in reverse order actually was (is?) an optimization: it ensures that GAP immediately allocates an object of the right size.
Now of course you (hopefully) get the same effect with `EmptyPlist`, but there are some cases where GAP (sadly) silentnly shrinks the result again. One of them is when `obj` is `true` or `false`: In that case, upon assigning `list[1]`, GAP will notice that the list only contains one item, which is a bool, and convert to a blist (and that blist will be small again, I think).
Unfortunately, there is no `EmptyBlist` command.
In any case, with your change, I think `ListWithIdenticalEntries(5,true)` produces a blist, while with the old code, it produced a plist:
```
gap> foo:=ListWithIdenticalEntries(5,true);
[ true, true, true, true, true ]
gap> TNUM_OBJ(foo);
[ 36, "list (plain,hom)" ]
```
That's probably fine, but something to be aware of.

There is another performance concern (theoretical, I have not verified it): GAP tries to compute whether a list is homogenous or not. Of course the lists we produce here always will be homogeneous, and even sorted. I am somewhat concerned that each time we append an element, we perform these tests...
We definitely do not perform them when inserting elements in reverse order, though, because then until the very end, `list[i]` is not set, and so the list cannot be homogenous, sorted, etc.

That said, I actually at one point wrote a C implementation of this function, but then failed to polish it into a PR...

Reply:
Take a look here: https://github.com/gap-system/gap/commit/0eb335dcfaeab1c2b5959e99f732caee50904e83

I just rebased and slightly cleaned up that old code. But it's still not very much tested. But perhaps you are interested in picking it up, and using it in this PR?

Reply:
Thanks @fingolfin, that's quite enlightening.

I had assumed that filling the list in from the end was to avoid calling `EmptyPlist`, but I didn't realise there would be other things to worry about, as you mentioned. Certainly we don't want GAP wasting effort every time we add the same entry to the list.

I'm happy to continue with this PR, and cherry-pick https://github.com/gap-system/gap/commit/0eb335dcfaeab1c2b5959e99f732caee50904e83 into this branch, but I'm not quite sure what sure what needs to be done.

Some questions:

* Shall I scrap the GAP method completely, and instead immediately call your C function `LIST_WITH_IDENTICAL_ENTRIES`?
* Your C function starts with the check `!IS_POS_INTOBJ(n)`, but we want to allow `0` - what is the correct way of checking this? (I'm not familiar with the names of GAP kernel stuff)
* This problem aside, do you consider the C function to be complete?

(In particular, I note the lines like `case T_INT:` and `case T_INTPOS:` etc, and I can't work out why they're there. I understand the rest, except for some of the blist code.)

Given answers to those questions, I'll be happy to update this PR with your C function, and I'll write lots of tests.

Reply:
I see `!IS_NONNEG_INTOBJ` is what I want.

Reply:
The `T_INT`, `T_INTPOS` business is to deal with the fact that we use tnums to remember certain things about lists. These record if the list is homogenous, or everything in it is a FFE, or a Cyclic. Also, we record if the lists are strictly sorted.

However, actually I think there is a mistake (or something extra going on?) because these lists aren't strictly sorted, as they contain the same value repeated several times!

Reply:
You could just use `T_PLIST`, this stuff just provides extra functionality which would otherwise be calculated later.

In general, if you are editing an existing `PLIST`, you should be careful to clear filters, else you can have a list which think it's sorted, and continues to think it is sorted after you have changed some values in it.

Reply:
Yes you're right, these lists aren't strictly sorted in general. I'll remove the two occurrences of `_SSORT` to give `T_PLIST_CYC` and `T_PLIST_HOM`, respectively.